### PR TITLE
Fix some bugs in HBO

### PR DESCRIPTION
**Pull request by pranjalssh**
Commit 1: Depends on https://github.com/prestodb/presto/pull/18340

Found some bugs in HBO which were leading to some optimizations not happening. Fixed them and dded a test which checks that we are able to optimize a query by changing its distribution type to BROADCAST.

Commit 2: Fix NPE in join canonicalization, where we called `accept(source).get()` without checking if optional is present. This happens when plan canonicalization is not implemented for some node.

Commit 3: 

* Fix bug in Memo/IterativeOptimizer, which assigned statsEquivalentPlanNode to GroupReference, which shouldn't happen. FIxed it by assigning to the group member instead
* Fixed HashGenerationOptimizer which was not propogating stats equivalent information when creating a Project node. 
* Added test described above

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/CanonicalPlanGenerator.java [hunk 1]
```diff
@@ -267,7 +267,14 @@ && shouldMergeJoinNodes(node.getType())) {
             sources.sort(comparing(sourceKeys::get));
         }
 
-        List<PlanNode> newSources = sources.stream().map(source -> source.accept(this, context).get()).collect(toImmutableList());
+        ImmutableList.Builder<PlanNode> newSources = ImmutableList.builder();
+        for (PlanNode source : sources) {
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/CanonicalPlanGenerator.java:271]: longer term, we'd want to support all node types, right?

**Review by rschlussel — COMMENTED** [review]
Make thrift compatible - LGTM
Fix join cannonicalization - can you add more details to the commit message about what the bug is?
Fixup some issues - can you add a more specific commit title and more details in the commit message about what you are fixing?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/CanonicalPlanGenerator.java (hunk 1 above)
Comment (pranjalssh) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/CanonicalPlanGenerator.java:271]: Yes, just focusing on productionizing it first

**Comment by pranjalssh** [comment]
Split 1st commit into different PR

**Review by rschlussel — APPROVED** [review]

**Pull request merged**